api/controllers/AssignmentController.js

```diff
@@ -119,14 +98,60 @@ module.exports = class AssignmentController extends classMixin(OrderedDataContro
       )
     }
 
-    if (judgement.grade > 0 && !parameterizedAssignment.dataValues.solved) {
-      await parameterizedAssignment.update({ solved: true })
-    }
-
     return res.status(200).send(
       {
         success: true, judgement
       }
     )
   }
+
+  async processAsyncJudge (req, res, aux, assignment, solution, parameterizedAssignment) {
+    let judgement = { grade: 0, msg: 'Wrong! Please try again.' }
+    const judge = new assignment.Judge(assignment.judge, req.user)
+
+    res.status(202).send({ code: 202, judgement: { grade: 0, msg: 'Solution is being processed. Please wait!' } })
+
+    try {
+      judgement = await judge.judge(aux, req.user, assignment.Judge, solution)
+      await this.updateSolution(req, parameterizedAssignment, solution)
+      await this.updateSolved(judgement, parameterizedAssignment)
+      appEmitter.emit('solution', { ...judgement })
+    } catch (err) {
+      logger.error(err)
+      judgement = { grade: 0, msg: err.message }
+
+      appEmitter.emit('solution', { ...judgement })
+    }
+  }
+
+  processSolution (req, res, aux, assignment, solution, parameterizedAssignment) {
+    switch (Object.getPrototypeOf(assignment.Judge.prototype.constructor).name) {
```

Comment:
Don't inspect the class name like this! Instead, use an attribute in the Judge instance to see if the Judge is sync or async, `assignment.Judge.isAsync`. This can be set to `false` by the `BaseJudge` and then overwritten to `true` by SolidityJudge. Do not use prototype here. That way, if we have other judges that want to be async, they can do so without requiring changes in this module.